### PR TITLE
CosmosDB support

### PR DESCRIPTION
**Pull request by jesty**
Hi Lukas,
I develop CosmosDB support. I hope that you like it.

### CODE REVIEW

File: providers/cosmosdb/shedlock-provider-cosmosdb/src/main/java/net/javacrumbs/shedlock/provider/cosmosdb/CosmosDBProvider.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ * <p>
+ * lockedAt and lockedBy are just for troubleshooting and are not read by the code
+ *
+ * <ol>
+ * <li>
+ * Attempts to insert a new lock record. As an optimization, we keep in-memory track of created lock records. If the record
+ * has been inserted, returns lock.
+ * </li>
+ * <li>
+ * We will invoke a stored procedure (checkLockAndAcquire.js) to check if the locke exists and update table.
```
Comment (lukas-krecan) [inline providers/cosmosdb/shedlock-provider-cosmosdb/src/main/java/net/javacrumbs/shedlock/provider/cosmosdb/CosmosDBProvider.java]: typo - locke

File: providers/cosmosdb/shedlock-provider-cosmosdb/src/main/java/net/javacrumbs/shedlock/provider/cosmosdb/CosmosDBProvider.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ * <li>
+ * When unlocking, lock_until is set to now.
+ * </li>
+ * </ol>
+ */
+public class CosmosDBProvider implements LockProvider {
+
+    public static final String ACQUIRE_LOCK_STORED_PROCEDURE = "acquireLock";
+    private final CosmosContainer container;
+    private String lockGroup;
```
Comment (lukas-krecan) [inline providers/cosmosdb/shedlock-provider-cosmosdb/src/main/java/net/javacrumbs/shedlock/provider/cosmosdb/CosmosDBProvider.java]: Can be final?

File: providers/cosmosdb/shedlock-provider-cosmosdb/src/main/java/net/javacrumbs/shedlock/provider/cosmosdb/CosmosDBProvider.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                .block();
+
+        if (!hasLock) {
+            return Optional.of(new CosmosDBLock(container, lockConfiguration, lockGroup, hostname));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    protected Object[] getProcedureParams(LockConfiguration lockConfiguration, long now) {
```
Comment (lukas-krecan) [inline providers/cosmosdb/shedlock-provider-cosmosdb/src/main/java/net/javacrumbs/shedlock/provider/cosmosdb/CosmosDBProvider.java]: Can be package protected?

**Comment by lukas-krecan** [comment]
I have checked CosmoDB and I am afraid that I can not support it because I can not test it. I do not have Windows and I do not want to pay for Azure. Feel free to create a separate project with this provider and I will be happy to reference it from the ShedLock README.

**Comment by jesty** [comment]
Hi @lukas-krecan,  I just created https://github.com/jesty/shedlock-provider-cosmosdb with the provider for CosmosDB. Thanks in advance if you can reference it from the ShedLock README.

**Comment by lukas-krecan** [comment]
Will do, thanks a lot

**Comment by lukas-krecan** [comment]
Link added

**Pull request closed without merging**